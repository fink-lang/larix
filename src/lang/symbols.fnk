{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
