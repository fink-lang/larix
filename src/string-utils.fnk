unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
