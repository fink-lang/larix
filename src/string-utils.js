import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

