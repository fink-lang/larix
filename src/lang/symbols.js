import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
