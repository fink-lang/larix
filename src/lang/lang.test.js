
import {other_token} from '@fink/prattler/symbols';
import {parse_expr, parse} from '..';
import {strip_block} from '../string-utils';


describe('foo', ()=> {
  it('parses', ()=> {
    expect(
      parse(strip_block`
        foo.ni(bar+spam)
        bar:: spam
      `)).toMatchSnapshot();
  });
});


