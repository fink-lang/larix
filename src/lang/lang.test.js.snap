// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foo parses 1`] = `
module (1:0-2:10)
  call (1:0-1:16)
    member . (1:0-1:6)
      other (1:0-1:3) foo
      other (1:4-1:6) ni
    arithm + (1:7-1:15)
      other (1:7-1:10) bar
      other (1:11-1:15) spam

  call (2:0-2:6)
    other (2:0-2:3) bar
    other (2:6-2:10) spam

`;
